proposals/ref-unsafe-in-iterators-async.md

```diff
@@ -43,11 +43,112 @@ lock (this)
+>
+> [...]
+>
 > ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).~~
-> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.
+> An iterator ~~block~~ **declaration ([§15.14][iterators])** always defines a safe context
+> **(its signature included),** even when the declaration is nested in an unsafe context,
+> **unless the iterator declaration is marked with the `unsafe` modifier ([§23.2][unsafe-contexts]).**
+
+This also means that a property or an indexer defines a safe context including its `set` accessor provided:
+- it is an iterator (because its `get` accessor is implemented via an iterator block) and
+- it does not have the `unsafe` modifier.
+
+If a method is `partial`, only the implementing declaration can be an iterator.
+
+For example:
+
+```cs
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class A : System.Attribute { }
+unsafe partial class C
+{ // unsafe context
+    [/* safe context */ A]
+    IEnumerable<int> M1(
+        /* safe context */)
+    { // safe context
+        yield return 1;
+    }
+    IEnumerable<int> M2()
+    { // safe context
+        unsafe
+        { // unsafe context
+            { // unsafe context
+                yield return 1; // error: `yield return` in unsafe context
+            }
+        }
+    }
+    [/* unsafe context */ A]
+    unsafe IEnumerable<int> M3(
+        /* unsafe context */)
+    { // unsafe context
+        yield return 1; // error: `yield return` in unsafe context
+    }
+    [/* safe context */ A]
+    IEnumerable<int> this[
+        /* safe context */ string x]
+    { // safe context
+        get
+        { // safe context
+            yield return 1;
+        }
+        set { /* safe context */ }
+    }
+    [/* unsafe context */ A]
+    unsafe IEnumerable<int> this[
+        /* unsafe context */ int x]
+    { // unsafe context
+        get
+        { // unsafe context
+            yield return 1; // error: `yield return` in unsafe context
+        }
+        set { /* unsafe context */ }
+    }
+    IEnumerable<int> M4()
+    {
+        yield return 1;
+        var lam1 = async () =>
+        { // safe context
+          // note: in Roslyn, this is an unsafe context in LangVersion 12 and lower
+            await Task.Yield();
+        };
+        unsafe
+        {
+            var lam2 = () =>
+            { // unsafe context, lambda cannot be an iterator
+                yield return 1; // error: yield cannot be used in lambda
+            };
+        }
+        async void local()
+        { // safe context
+          // note: in Roslyn, this is an unsafe context in LangVersion 12 and lower
+            await Task.Yield();
+        }
+        local();
+    }
+    partial IEnumerable<int> M5(); // unsafe context, no iterator body
```

Comment:
>unsafe [](http://example.com/codeflow?start=38&length=6)

Does that mean an explicit `unsafe` will be allowed on the definition part only for iterators with language version 13?
```csharp
partial IEnumerable<int> M6();
partial unsafe IEnumerable<int> M6() { yield break; }
```

Reply:
Yes, exactly.

Reply:
Shouldn't the mismatch (of safe/unsafe contexts for the partial method signatures) result in an error?